File: src/ados/api/routes/logs.py
"""Log viewing routes, sourced from the local logging and telemetry store.

These two endpoints keep the legacy response shape that older Mission Control
clients expect, but they no longer hold their own copy of the logs. The agent's
log records flow to the durable on-disk store over the structlog socket handler
(``core/logd_ship.py``); these endpoints read them back from the store's query
API over its trusted local Unix socket.

* ``GET /api/logs`` — recent entries, mapped from the store's ``/v1/query`` rows
  to the legacy ``{ timestamp, level, logger, message }`` tuple. If the store is
  not reachable the endpoint degrades to an empty list with a warning rather
  than a 500: losing history degrades debugging, not flight.
* ``GET /api/logs/stream`` — a Server-Sent-Events stream proxied from the
  store's ``/v1/tail`` and re-mapped to the legacy SSE frame shape.

The store survives reboots and is reachable when the network is down, which is
exactly when an in-memory buffer was least useful. The stderr/journald sink
remains the always-on primary; the store is the durable secondary that these
endpoints read.
"""

from __future__ import annotations

import json
from datetime import datetime, timezone
from typing import Any

import httpx
from fastapi import APIRouter, Body, Query
from fastapi.concurrency import run_in_threadpool
from fastapi.responses import JSONResponse, StreamingResponse

from ados.core.logging import get_logger
from ados.core.paths import LOGD_QUERY_SOCK

log = get_logger("api.logs")

router = APIRouter()

# The store's query-API base over the trusted Unix socket. The host portion is a
# placeholder httpx requires; the UDS transport routes to the socket regardless.
_UPSTREAM_BASE = "http://logd"

# Hard ceiling on the page size the legacy surface will request from the store.
# The legacy clients ask for tens of entries; this caps a pathological request.
_MAX_LIMIT = 1000

# Connect fast so a missing store degrades the endpoint at once; the read leg is
# bounded because the legacy query is a single bounded page.
_QUERY_TIMEOUT = httpx.Timeout(connect=2.0, read=10.0, write=5.0, pool=2.0)

# The live tail has no read-idle ceiling (the store sends keep-alive comment
# frames); the connect leg still fails fast on a missing store.
_TAIL_TIMEOUT = httpx.Timeout(connect=2.0, read=None, write=5.0, pool=2.0)

# Module-level singletons over the UDS, created on first use. Separate clients so
# the unbounded-read tail timeout never leaks onto the bounded query. Tests can
# override either by assignment (e.g. an ``httpx.MockTransport``-backed client).
_query_client: httpx.AsyncClient | None = None
_tail_client: httpx.AsyncClient | None = None


def _get_query_client() -> httpx.AsyncClient:
    global _query_client
    if _query_client is None:
        transport = httpx.AsyncHTTPTransport(uds=str(LOGD_QUERY_SOCK))
        _query_client = httpx.AsyncClient(
            base_url=_UPSTREAM_BASE, transport=transport, timeout=_QUERY_TIMEOUT
        )
    return _query_client


def _get_tail_client() -> httpx.AsyncClient:
    global _tail_client
    if _tail_client is None:
        transport = httpx.AsyncHTTPTransport(uds=str(LOGD_QUERY_SOCK))
        _tail_client = httpx.AsyncClient(
            base_url=_UPSTREAM_BASE, transport=transport, timeout=_TAIL_TIMEOUT
        )
    return _tail_client


async def aclose_clients() -> None:
    """Close the shared upstream clients. Called on app shutdown."""
    global _query_client, _tail_client
    if _query_client is not None:
        await _query_client.aclose()
        _query_client = None
    if _tail_client is not None:
        await _tail_client.aclose()
        _tail_client = None


def _legacy_entry(row: dict[str, Any]) -> dict[str, Any]:
    """Map one store log row onto the legacy ``/api/logs`` entry shape.

    The store row is ``{ id, ts_us, session, source, level, target, msg,
    fields }``; the legacy consumer expects ``{ seq, timestamp, level, logger,
    message }`` with an ISO-8601 timestamp and an upper-case level name.
    """
    ts_us = row.get("ts_us")
    timestamp: str
    if isinstance(ts_us, (int, float)):
        timestamp = datetime.fromtimestamp(
            ts_us / 1_000_000, tz=timezone.utc
        ).isoformat()
    else:
        timestamp = datetime.now(tz=timezone.utc).isoformat()
    return {
        "seq": row.get("id"),
        "timestamp": timestamp,
        "level": str(row.get("level", "")).upper(),
        "logger": row.get("target") or row.get("source") or "",
        "message": row.get("msg", ""),
    }


@router.get("/logs")
async def get_logs(
    level: str | None = Query(None),
    service: str | None = Query(None),
    limit: int = Query(50, ge=1, le=_MAX_LIMIT),
    offset: int = Query(0, ge=0),
):
    """Recent log entries, sourced from the durable store.

    Maps the store's rows to the legacy response shape so existing clients keep
    working. If the store is unreachable the endpoint returns an empty list with
    a ``warning`` field instead of a 500 — history is observability, not flight.
    """
    # Ask the store for enough rows to satisfy the offset window, then page in
    # Python so the legacy offset/limit contract is honored without leaking the
    # store's keyset-cursor model to the legacy caller.
    want = min(offset + limit, _MAX_LIMIT)
    params: dict[str, Any] = {"kind": "logs", "limit": want}
    if level:
        params["level"] = level.lower()
    if service:
        params["source"] = service

    client = _get_query_client()
    try:
        resp = await client.get("/v1/query", params=params)
    except (httpx.ConnectError, httpx.ConnectTimeout, FileNotFoundError, OSError):
        log.warning("logs_store_unreachable")
        return JSONResponse(
            content={
                "entries": [],
                "total": 0,
                "limit": limit,
                "offset": offset,
                "warning": "logging store unavailable",
            }
        )
    except httpx.HTTPError as exc:
        log.warning("logs_store_error", error=str(exc))
        return JSONResponse(
            content={
                "entries": [],
                "total": 0,
                "limit": limit,
                "offset": offset,
                "warning": "logging store query failed",
            }
        )

    if resp.status_code >= 400:
        log.warning("logs_store_status", status=resp.status_code)
        return JSONResponse(
            content={
                "entries": [],
                "total": 0,
                "limit": limit,
                "offset": offset,
                "warning": f"logging store returned {resp.status_code}",
            }
        )

    try:
        body = resp.json()
    except ValueError:
        return JSONResponse(
            content={
                "entries": [],
                "total": 0,
                "limit": limit,
                "offset": offset,
                "warning": "logging store response was not JSON",
            }
        )

    rows = body.get("data") if isinstance(body, dict) else None
    if not isinstance(rows, list):
        rows = []

    # The store returns newest-first; the service filter is already applied
    # store-side via ``source``, but it matches a source prefix loosely there,
    # so re-apply the legacy substring semantics here for parity.
    mapped = [_legacy_entry(r) for r in rows if isinstance(r, dict)]
    if service:
        mapped = [e for e in mapped if service in e.get("logger", "")]
    if level:
        level_upper = level.upper()
        mapped = [e for e in mapped if e["level"] == level_upper]

    total = len(mapped)
    window = mapped[offset : offset + limit]
    return {
        "entries": window,
        "total": total,
        "limit": limit,
        "offset": offset,
    }


@router.get("/logs/stream")
async def stream_logs(
    level: str | None = Query(None),
    service: str | None = Query(None),
):
    """Server-Sent Events stream proxied from the store's live tail.

    Each store tail row is re-mapped to the legacy ``data: <json>`` frame so
    existing EventSource clients keep working. A replay of the most recent
    entries is requested so a fresh stream shows recent context, matching the
    old snapshot behavior. Keep-alive comment frames pass through. If the store
    is unreachable the stream closes cleanly so the client reconnects.
    """
    params: dict[str, Any] = {"kind": "logs", "replay": 100}
    if level:
        params["level"] = level.lower()
    if service:
        params["source"] = service

    client = _get_tail_client()

    async def gen():
        upstream_req = client.build_request("GET", "/v1/tail", params=params)
        try:
            upstream = await client.send(upstream_req, stream=True)
        except (httpx.ConnectError, httpx.ConnectTimeout, FileNotFoundError, OSError):
            log.warning("logs_stream_store_unreachable")
            yield ": logging store unavailable\n\n"
            return
        except httpx.HTTPError as exc:
            log.warning("logs_stream_store_error", error=str(exc))
            yield ": logging store stream failed\n\n"
            return

        try:
            if upstream.status_code >= 400:
                yield ": logging store stream error\n\n"
                return
            async for line in upstream.aiter_lines():
                if not line:
                    continue
                if line.startswith(":"):
                    # Pass keep-alive / notice comment frames straight through.
                    yield f"{line}\n\n"
                    continue
                if not line.startswith("data:"):
                    continue
                payload = line[len("data:") :].strip()
                if not payload:
                    continue
                try:
                    row = json.loads(payload)
                except json.JSONDecodeError:
                    continue
                # The store may publish a "lagged" notice frame on the tail;
                # forward it as a comment so the legacy client does not try to
                # render it as a log entry.
                if isinstance(row, dict) and row.get("kind") == "lagged":
                    yield ": tail lagged\n\n"
                    continue
                if not isinstance(row, dict):
                    continue
                yield f"data: {json.dumps(_legacy_entry(row))}\n\n"
        finally:
            # Releases the upstream connection on completion AND on a client
            # disconnect (Starlette cancels the generator).
            await upstream.aclose()

    return StreamingResponse(gen(), media_type="text/event-stream")


@router.post("/logs/push")
async def push_logs(
    body: dict[str, Any] = Body(default_factory=dict),  # noqa: B008 - FastAPI body marker
):
    """Request an explicit cloud export of a chosen log window.

    This is a thin trigger. It records the operator's window selector and lets
    the cloud service do the export, upload, and mark-synced — that service owns
    the trusted store socket and the cloud client. This endpoint never reads the
    store, uploads bytes, or marks rows itself.

    The agent's existing ``X-ADOS-Key`` auth middleware gates this route, so a
    paired agent requires the key (and a same-origin browser is trusted) before
    the request reaches here. The cloud service is what refuses the actual push
    when the agent is in local mode, is not cloud-paired, or has cloud log push
    disabled; an agent with nothing to sync logging fully to disk is correct.

    Request body (all optional): ``{ session?: int, since?: str,
    kinds?: [str] }``. ``kinds`` is any subset of ``logs``, ``metrics``,
    ``events``, ``hw``; empty or absent means all four. ``since`` accepts a
    relative ``-5m``/``-2h``/``-1d``, an epoch-microsecond integer, or an ISO
    timestamp. ``wait`` (default true) controls whether the call blocks briefly
    for the result; set it false to get a 202 as soon as the request is
    recorded.

    Responses:

    * ``200`` with ``{ accepted, request_id, pushed, deduped, bytes, rows,
      synced, error, pending, ... }`` when the result lands (or the brief poll
      window closes with the push still pending);
    * ``202`` with the same shape (``pending: true``) when ``wait`` is false;
    * ``400`` ``{ error: { code, message } }`` on a malformed selector.
    """
    from ados.services.cloud.log_push_trigger import (
        LogPushTriggerError,
        build_request,
        trigger_push,
    )

    session = body.get("session")
    since = body.get("since")
    raw_kinds = body.get("kinds")
    wait = body.get("wait", True)

    kinds: list[str] | None
    if raw_kinds is None:
        kinds = None
    elif isinstance(raw_kinds, str):
        kinds = [k.strip() for k in raw_kinds.split(",") if k.strip()]
    elif isinstance(raw_kinds, list):
        kinds = [str(k) for k in raw_kinds]
    else:
        return JSONResponse(
            status_code=400,
            content={"error": {"code": "bad_kinds", "message": "kinds must be a list or comma string"}},
        )

    if session is not None and not isinstance(session, int):
        return JSONResponse(
            status_code=400,
            content={"error": {"code": "bad_session", "message": "session must be an integer"}},
        )

    try:
        request = build_request(
            session=session,
            since=str(since) if since is not None else None,
            kinds=kinds,
        )
        # The trigger does blocking file I/O and a short polling sleep; run it
        # off the event loop so a slow result poll does not stall the server.
        result = await run_in_threadpool(trigger_push, request, wait=bool(wait))
    except LogPushTriggerError as exc:
        log.warning("logs_push_rejected", code=exc.code)
        return JSONResponse(
            status_code=400,
            content={"error": {"code": exc.code, "message": exc.message}},
        )

    status = 202 if result.get("pending") else 200
    return JSONResponse(status_code=status, content=result)


__all__ = ["router", "aclose_clients"]

File: crates/ados-control/src/routes/logs.rs
//! `GET /api/logs` — recent log entries from the durable store, in the
//! `{seq, timestamp, level, logger, message}` entry shape the GCS log tier and
//! the relayed-read path consume.
//!
//! The store answers newest first. The route asks for enough rows to cover the
//! `offset` window (capped at [`MAX_LIMIT`]) and pages here, so the caller keeps
//! its offset/limit contract. An unreachable or failing store is an empty page
//! with a `warning` naming why, not a `500`: log history is observability, not
//! flight.

use axum::extract::Query;
use axum::response::{IntoResponse, Response};
use axum::Json;
use serde::Deserialize;
use serde_json::{json, Value};
use time::format_description::well_known::Rfc3339;
use time::OffsetDateTime;

use crate::ipc::logd_client::LogdQueryClient;

/// Hard ceiling on the page the route will request from the store.
pub const MAX_LIMIT: i64 = 1000;

#[derive(Debug, Deserialize)]
pub struct LogsQuery {
    #[serde(default)]
    pub level: Option<String>,
    #[serde(default)]
    pub service: Option<String>,
    #[serde(default = "default_limit")]
    pub limit: i64,
    #[serde(default)]
    pub offset: i64,
}

fn default_limit() -> i64 {
    50
}

/// `GET /api/logs?level=&service=&limit=&offset=`.
pub async fn get_logs(Query(q): Query<LogsQuery>) -> Response {
    logs_from(&LogdQueryClient::default_socket(), q).await
}

async fn logs_from(client: &LogdQueryClient, q: LogsQuery) -> Response {
    if !(1..=MAX_LIMIT).contains(&q.limit) || q.offset < 0 {
        return crate::routes::detail(
            axum::http::StatusCode::UNPROCESSABLE_ENTITY,
            format!("limit must be 1..={MAX_LIMIT} and offset must be >= 0"),
        );
    }
    let level = q.level.as_deref().filter(|l| !l.is_empty());
    let service = q.service.as_deref().filter(|s| !s.is_empty());
    let want = (q.offset + q.limit).min(MAX_LIMIT);
    let mut params: Vec<(&str, String)> =
        vec![("kind", "logs".to_string()), ("limit", want.to_string())];
    if let Some(l) = level {
        params.push(("level", l.to_ascii_lowercase()));
    }
    if let Some(s) = service {
        params.push(("source", s.to_string()));
    }
    let path = format!(
        "/v1/query?{}",
        crate::ipc::logd_client::encode_query(&params)
    );
    let empty = |warning: String| {
        Json(json!({
            "entries": [],
            "total": 0,
            "limit": q.limit,
            "offset": q.offset,
            "warning": warning,
        }))
        .into_response()
    };
    let (status, body) = match client.get(&path).await {
        Ok(r) => r,
        Err(_) => return empty("logging store unavailable".to_string()),
    };
    if status >= 400 {
        return empty(format!("logging store returned {status}"));
    }
    let Ok(parsed) = serde_json::from_slice::<Value>(&body) else {
        return empty("logging store response was not JSON".to_string());
    };
    let rows = parsed
        .get("data")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    let level_upper = level.map(str::to_ascii_uppercase);
    let entries: Vec<Value> = rows
        .iter()
        .filter_map(Value::as_object)
        .map(legacy_entry)
        .filter(|e| service.is_none_or(|s| e["logger"].as_str().unwrap_or("").contains(s)))
        .filter(|e| {
            level_upper
                .as_deref()
                .is_none_or(|l| e["level"].as_str() == Some(l))
        })
        .collect();
    let total = entries.len();
    let window: Vec<Value> = entries
        .into_iter()
        .skip(q.offset as usize)
        .take(q.limit as usize)
        .collect();
    Json(json!({
        "entries": window,
        "total": total,
        "limit": q.limit,
        "offset": q.offset,
    }))
    .into_response()
}

/// One store row (`{id, ts_us, source, level, target, msg, ..}`) as a log entry.
/// A row without a timestamp keeps `timestamp: null`: the time it was logged is
/// unknown, and the time it was read is not a substitute.
fn legacy_entry(row: &serde_json::Map<String, Value>) -> Value {
    let timestamp = row
        .get("ts_us")
        .and_then(Value::as_i64)
        .and_then(|us| OffsetDateTime::from_unix_timestamp_nanos(i128::from(us) * 1_000).ok())
        .and_then(|t| t.format(&Rfc3339).ok());
    let text = |k: &str| row.get(k).and_then(Value::as_str).filter(|s| !s.is_empty());
    json!({
        "seq": row.get("id").cloned().unwrap_or(Value::Null),
        "timestamp": timestamp,
        "level": text("level").unwrap_or("").to_ascii_uppercase(),
        "logger": text("target").or_else(|| text("source")).unwrap_or(""),
        "message": text("msg").unwrap_or(""),
    })
}

/// The `GET /api/logs/stream` filters.
#[derive(Debug, Deserialize)]
pub struct StreamQuery {
    #[serde(default)]
    pub level: Option<String>,
    #[serde(default)]
    pub service: Option<String>,
}

/// Rows the live tail replays first, so a fresh stream shows recent context.
const TAIL_REPLAY: u32 = 100;

/// How long the store may take to answer the tail's head.
const TAIL_HEAD_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(5);

/// `GET /api/logs/stream` — Server-Sent Events from the store's live tail,
/// each row re-shaped to the log-entry form. Keep-alive comment frames pass
/// through; an unreachable store ends the stream with a comment so the
/// EventSource reconnects.
pub async fn get_logs_stream(Query(q): Query<StreamQuery>) -> Response {
    let (mut tx, body) = http_body_util::channel::Channel::<bytes::Bytes>::new(16);
    let socket = crate::ipc::logd_client::default_logd_socket();
    tokio::spawn(async move { pump_tail(&socket, &q, &mut tx).await });
    axum::http::Response::builder()
        .header(axum::http::header::CONTENT_TYPE, "text/event-stream")
        .header(axum::http::header::CACHE_CONTROL, "no-cache")
        .body(axum::body::Body::new(body))
        .map_or_else(
            |_| axum::http::StatusCode::INTERNAL_SERVER_ERROR.into_response(),
            IntoResponse::into_response,
        )
}

async fn pump_tail(
    socket: &std::path::Path,
    q: &StreamQuery,
    tx: &mut http_body_util::channel::Sender<bytes::Bytes>,
) {
    use http_body_util::BodyExt;

    let mut params: Vec<(&str, String)> = vec![
        ("kind", "logs".to_string()),
        ("replay", TAIL_REPLAY.to_string()),
    ];
    if let Some(l) = q.level.as_deref().filter(|l| !l.is_empty()) {
        params.push(("level", l.to_ascii_lowercase()));
    }
    if let Some(s) = q.service.as_deref().filter(|s| !s.is_empty()) {
        params.push(("source", s.to_string()));
    }
    let path = format!(
        "/v1/tail?{}",
        crate::ipc::logd_client::encode_query(&params)
    );
    let Some(mut body) = open_tail(socket, &path).await else {
        let _ = tx
            .send_data(bytes::Bytes::from_static(
                b": logging store unavailable\n\n",
            ))
            .await;
        return;
    };
    let mut pending: Vec<u8> = Vec::new();
    while let Some(frame) = body.frame().await {
        let Ok(frame) = frame else { return };
        let Ok(data) = frame.into_data() else {
            continue;
        };
        pending.extend_from_slice(&data);
        while let Some(end) = pending.iter().position(|b| *b == b'\n') {
            let line: Vec<u8> = pending.drain(..=end).collect();
            let line = String::from_utf8_lossy(&line);
            if let Some(out) = sse_frame(line.trim_end_matches(['\r', '\n'])) {
                // The client went away: dropping the upstream body closes the tail.
                if tx.send_data(bytes::Bytes::from(out)).await.is_err() {
                    return;
                }
            }
        }
    }
}

/// Open the store's tail over its Unix socket: the streaming response body, or
/// `None` when the store is absent or refuses.
async fn open_tail(socket: &std::path::Path, path: &str) -> Option<hyper::body::Incoming> {
    let stream = tokio::net::UnixStream::connect(socket).await.ok()?;
    let (mut sender, conn) =
        hyper::client::conn::http1::handshake(hyper_util::rt::TokioIo::new(stream))
            .await
            .ok()?;
    tokio::spawn(async move {
        let _ = conn.await;
    });
    let request = axum::http::Request::get(path)
        .header(axum::http::header::HOST, "logd")
        .body(http_body_util::Empty::<bytes::Bytes>::new())
        .ok()?;
    let resp = tokio::time::timeout(TAIL_HEAD_TIMEOUT, sender.send_request(request))
        .await
        .ok()?
        .ok()?;
    resp.status().is_success().then(|| resp.into_body())
}

/// One upstream SSE line as the downstream frame, or `None` to drop it.
fn sse_frame(line: &str) -> Option<String> {
    if line.starts_with(':') {
        return Some(format!("{line}\n\n"));
    }
    let payload = line.strip_prefix("data:")?.trim();
    let row: Value = serde_json::from_str(payload).ok()?;
    let row = row.as_object()?;
    if row.get("kind").and_then(Value::as_str) == Some("lagged") {
        return Some(": tail lagged\n\n".to_string());
    }
    Some(format!("data: {}\n\n", legacy_entry(row)))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn row(id: i64, level: &str, target: &str, msg: &str) -> Value {
        json!({"id": id, "ts_us": 1_700_000_000_000_000i64 + id, "level": level, "target": target, "source": "src", "msg": msg})
    }

    #[test]
    fn a_store_row_maps_to_the_entry_shape() {
        let e = legacy_entry(row(7, "warn", "ados_video", "hi").as_object().unwrap());
        assert_eq!(e["seq"], 7);
        assert_eq!(e["level"], "WARN");
        assert_eq!(e["logger"], "ados_video");
        assert_eq!(e["message"], "hi");
        assert!(e["timestamp"]
            .as_str()
            .unwrap()
            .starts_with("2023-11-14T22:13:20"));
    }

    #[test]
    fn a_row_without_a_timestamp_is_not_stamped_now() {
        let e = legacy_entry(json!({"id": 1, "msg": "x"}).as_object().unwrap());
        assert!(e["timestamp"].is_null());
    }

    #[tokio::test]
    async fn an_unreachable_store_is_an_empty_page_with_a_warning() {
        let dir = tempfile::tempdir().unwrap();
        let client = LogdQueryClient::new(dir.path().join("absent.sock"));
        let q = LogsQuery {
            level: None,
            service: None,
            limit: 50,
            offset: 0,
        };
        let resp = logs_from(&client, q).await;
        assert_eq!(resp.status(), axum::http::StatusCode::OK);
        let body = axum::body::to_bytes(resp.into_body(), usize::MAX)
            .await
            .unwrap();
        let v: Value = serde_json::from_slice(&body).unwrap();
        assert_eq!(v["entries"], json!([]));
        assert_eq!(v["warning"], "logging store unavailable");
    }

    #[test]
    fn tail_lines_become_entry_frames_and_notices_become_comments() {
        assert_eq!(
            sse_frame(": keep-alive").as_deref(),
            Some(": keep-alive\n\n")
        );
        assert_eq!(
            sse_frame(r#"data: {"kind":"lagged","skipped":4}"#).as_deref(),
            Some(": tail lagged\n\n")
        );
        let frame = sse_frame(&format!("data: {}", row(3, "info", "ados_radio", "up"))).unwrap();
        let v: Value = serde_json::from_str(frame.trim().strip_prefix("data: ").unwrap()).unwrap();
        assert_eq!(v["logger"], "ados_radio");
        assert_eq!(v["level"], "INFO");
        assert!(sse_frame("event: x").is_none());
        assert!(sse_frame("data: not json").is_none());
    }
}
